src/transactions/ClawbackClaimableBalanceOpFrame.cpp

```diff
@@ -0,0 +1,105 @@
+// Copyright 2021 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "transactions/ClawbackClaimableBalanceOpFrame.h"
+#include "ledger/LedgerTxn.h"
+#include "transactions/SponsorshipUtils.h"
+#include "transactions/TransactionUtils.h"
+#include <Tracy.hpp>
+
+namespace stellar
+{
+
+ClawbackClaimableBalanceOpFrame::ClawbackClaimableBalanceOpFrame(
+    Operation const& op, OperationResult& res, TransactionFrame& parentTx)
+    : OperationFrame(op, res, parentTx)
+    , mClawbackClaimableBalance(mOperation.body.clawbackClaimableBalanceOp())
+{
+}
+
+bool
+ClawbackClaimableBalanceOpFrame::isVersionSupported(
+    uint32_t protocolVersion) const
+{
+    return protocolVersion >= 16;
+}
+
+bool
+ClawbackClaimableBalanceOpFrame::doApply(AbstractLedgerTxn& ltx)
+{
+    ZoneNamedN(applyZone, "ClawbackClaimableBalanceOp apply", true);
+
+    auto claimableBalanceLtxEntry =
+        stellar::loadClaimableBalance(ltx, mClawbackClaimableBalance.balanceID);
+    if (!claimableBalanceLtxEntry)
+    {
+        innerResult().code(CLAWBACK_CLAIMABLE_BALANCE_DOES_NOT_EXIST);
+        return false;
+    }
+
+    auto const& claimableBalance =
+        claimableBalanceLtxEntry.current().data.claimableBalance();
+
+    if (claimableBalance.asset.type() == ASSET_TYPE_NATIVE)
+    {
+        innerResult().code(CLAWBACK_CLAIMABLE_BALANCE_NOT_ISSUER);
+        return false;
+    }
+
+    switch (claimableBalance.asset.type())
+    {
+    case ASSET_TYPE_NATIVE:
+        innerResult().code(CLAWBACK_CLAIMABLE_BALANCE_NOT_ISSUER);
+        return false;
+
+    case ASSET_TYPE_CREDIT_ALPHANUM4:
+        if (!(claimableBalance.asset.alphaNum4().issuer == getSourceID()))
+        {
+            innerResult().code(CLAWBACK_CLAIMABLE_BALANCE_NOT_ISSUER);
+            return false;
+        }
+        break;
+    case ASSET_TYPE_CREDIT_ALPHANUM12:
+        if (!(claimableBalance.asset.alphaNum12().issuer == getSourceID()))
+        {
+            innerResult().code(CLAWBACK_CLAIMABLE_BALANCE_NOT_ISSUER);
+            return false;
+        }
+        break;
+    default:
+        throw std::runtime_error(
+            "ClawbackClaimableBalanceOpFrame - Unknown asset type");
+    }
+
+    if (!isClawbackEnabledOnClaimableBalance(
+            claimableBalanceLtxEntry.current()))
+    {
+        innerResult().code(CLAWBACK_CLAIMABLE_BALANCE_NOT_CLAWBACK_ENABLED);
+        return false;
+    }
+
+    auto header = ltx.loadHeader();
+    auto sourceAccount = loadSourceAccount(ltx, header);
+    removeEntryWithPossibleSponsorship(
```

Comment:
That's a good point. It can definitely be improved to accommodate claimable balances. I'd say I'm at least partially to blame here since I adapted it to work for claimable balances instead of improving the interface...

Reply:
Tracked in https://github.com/stellar/stellar-core/issues/2899.